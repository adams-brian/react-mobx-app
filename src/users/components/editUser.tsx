import { inject, observer } from 'mobx-react';
import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { Link } from 'react-router-dom';

import UsersState, { IUser } from '../state';

interface IEditUserProps extends RouteComponentProps<{id: string}> {
  usersState: UsersState;
}

export class EditUser extends React.Component<IEditUserProps, IUser> {
  constructor(props: IEditUserProps) {
    super(props);
    this.state = this.props.usersState.users.find(u => u._id === props.match.params.id) || 
      { _id: '', firstname: '', lastname: '' } as IUser;
  }

  public render() {
    return (
      <div className="edit-user">
        <h1>{this.state._id && this.state._id.length > 0 ? 'Edit' : 'Create'} User</h1>
        <form>
          <div className="form-group">
            <label htmlFor="firstname">Firstname</label>
            <input
              placeholder="Firstname"
              type="text"
              className="form-control"
              name="firstname"
              id="firstname"
              value={this.state.firstname}
              onChange={this.inputChanged}
              required={true}
            />
          </div>
          <div className="form-group">
            <label htmlFor="lastname">Lastname</label>
            <input
              placeholder="Lastname"
              type="text"
              className="form-control"
              name="lastname"
              id="lastname"
              value={this.state.lastname}
              onChange={this.inputChanged}
              required={true}
            />
          </div>
          <div>
            <Link to="/users">
              <button type="button" className="btn btn-secondary">Cancel</button>
            </Link>
            {' '}
            <button 
              className="btn btn-primary" 
              type="submit" 
              onClick={this.submit} 
              disabled={this.state.firstname.length === 0 || this.state.lastname.length === 0}
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }

  private inputChanged = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({
      [e.currentTarget.name]: e.currentTarget.value
    } as Pick<IUser, keyof IUser>);
  }

  private submit = (e: React.FormEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (this.state._id && this.state._id.length > 0) {
      this.props.usersState.updateUser(this.state);
    }
    else {
      this.props.usersState.createUser(this.state);
    }
    this.props.history.push('/users');
  }
}

export default inject('usersState')(
  observer(
    EditUser
  )
);
