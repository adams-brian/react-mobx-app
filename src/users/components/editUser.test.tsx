import { shallow } from 'enzyme';
import { createLocation, createMemoryHistory, Location, MemoryHistory } from 'history'
import * as React from 'react';
import * as sinon from 'sinon';

import UsersState from '../state';
import { EditUser } from './editUser';

interface IMatch {
  params: {id: string};
  isExact: boolean;
  path: string;
  url: string;
}

let element: JSX.Element;
let history: MemoryHistory;
let location: Location;
let match: IMatch;
let usersState: UsersState;

beforeEach(() => {
  usersState = new UsersState({
    createUser: sinon.fake(),
    deleteUser: sinon.fake(),
    loadUsers: sinon.fake.resolves([
      {
        _id: '123',
        firstname: 'abc',
        lastname: 'def'
      },
      {
        _id: '456',
        firstname: 'hij',
        lastname: 'klm'
      },
      {
        _id: '789',
        firstname: 'nop',
        lastname: 'qrs'
      }
    ]),
    updateUser: sinon.fake()
  });
  sinon.spy(usersState, 'createUser');
  sinon.spy(usersState, 'updateUser');
  return usersState.loadUsers();
})

describe("editUser", () => {

  describe("create user mode", () => {

    beforeEach(() => {
      history = createMemoryHistory();
      sinon.spy(history, 'push');
      location = createLocation('');
      match = {
        isExact: true,
        params: { id: '' },
        path: '',
        url: ''
      }
    
      element = (
        <EditUser
          history={history}
          location={location}
          match={match}
          staticContext={undefined}
          usersState={usersState}
        />
      )
    });

    it('renders as expected', () => {
      const component = shallow(element);
      expect(component).toMatchSnapshot();
    });
    it('calls createUser on the state', () => {
      const component = shallow(element);
      expect((usersState.createUser as sinon.SinonSpy).called).toBe(false);
      expect((history.push as sinon.SinonSpy).called).toBe(false);
      component.find('#firstname').first().simulate('change', { currentTarget: { name: 'firstname', value: 'new' } });
      component.find('#lastname').first().simulate('change', { currentTarget: { name: 'lastname', value: 'user' } });
      component.find('button[type="submit"]').first().simulate('click', { preventDefault: () => { /* noop */ }});
      expect((usersState.createUser as sinon.SinonSpy).calledOnceWithExactly({ _id: '', firstname: 'new', lastname: 'user'})).toBe(true);
      expect((history.push as sinon.SinonSpy).calledOnceWithExactly('/users')).toBe(true);
    });
  });

  describe("edit user mode", () => {

    beforeEach(() => {
      history = createMemoryHistory();
      sinon.spy(history, 'push');
      location = createLocation('');
      match = {
        isExact: true,
        params: { id: '456' },
        path: '',
        url: ''
      }
    
      element = (
        <EditUser
          history={history}
          location={location}
          match={match}
          staticContext={undefined}
          usersState={usersState}
        />
      )
    });

    it('renders as expected', () => {
      const component = shallow(element);
      expect(component).toMatchSnapshot();
    });
    it('calls updateUser on the state', () => {
      const component = shallow(element);
      expect((usersState.updateUser as sinon.SinonSpy).called).toBe(false);
      expect((history.push as sinon.SinonSpy).called).toBe(false);
      component.find('#firstname').first().simulate('change', { currentTarget: { name: 'firstname', value: 'updated' } });
      component.find('#lastname').first().simulate('change', { currentTarget: { name: 'lastname', value: 'user' } });
      component.find('button[type="submit"]').first().simulate('click', { preventDefault: () => { /* noop */ }});
      expect((usersState.updateUser as sinon.SinonSpy).calledOnceWithExactly({ _id: '456', firstname: 'updated', lastname: 'user'})).toBe(true);
      expect((history.push as sinon.SinonSpy).calledOnceWithExactly('/users')).toBe(true);
    });
  });
});
