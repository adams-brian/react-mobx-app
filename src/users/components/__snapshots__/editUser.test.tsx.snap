// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUser create user mode renders as expected 1`] = `
<div
  className="edit-user"
>
  <h1>
    Create
     User
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="firstname"
      >
        Firstname
      </label>
      <input
        className="form-control"
        id="firstname"
        name="firstname"
        onChange={[Function]}
        placeholder="Firstname"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastname"
      >
        Lastname
      </label>
      <input
        className="form-control"
        id="lastname"
        name="lastname"
        onChange={[Function]}
        placeholder="Lastname"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      <Link
        replace={false}
        to="/users"
      >
        <button
          className="btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </Link>
       
      <button
        className="btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`EditUser edit user mode renders as expected 1`] = `
<div
  className="edit-user"
>
  <h1>
    Edit
     User
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="firstname"
      >
        Firstname
      </label>
      <input
        className="form-control"
        id="firstname"
        name="firstname"
        onChange={[Function]}
        placeholder="Firstname"
        required={true}
        type="text"
        value="hij"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastname"
      >
        Lastname
      </label>
      <input
        className="form-control"
        id="lastname"
        name="lastname"
        onChange={[Function]}
        placeholder="Lastname"
        required={true}
        type="text"
        value="klm"
      />
    </div>
    <div>
      <Link
        replace={false}
        to="/users"
      >
        <button
          className="btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </Link>
       
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
